cdap-watchdog/src/main/java/io/cdap/cdap/metrics/collect/Distribution.java

```diff
@@ -0,0 +1,136 @@
+ */
+
+package io.cdap.cdap.metrics.collect;
+
+import io.cdap.cdap.api.metrics.MetricValue;
+
+import java.util.Arrays;
+
+/**
+ * Class is not thread-safe. It is the caller's responsibility to ensure thread safety.
+ */
+public class Distribution {
+    /**
+     * Trying to minimize memory usage for the buckets. Using a long[64] will always store 64 numbers,
+     * most of them being 0. A HashMap is an alternative, but boxing overhed is high (roughly 10X) for
+     * each long.
+     *
+     * A third alternative used is to increase array size till the highest bucket. But this involves
+     * repeatedly copying the array.
+     *
+     * The current approach strikes a balance by using an array of arrays. Ex. bucket 52 is stored in
+     * the 6th row, 5th column (6*8 + 5th column). Initially all rows are null. The moment a bucket increases
+     * by 1, a long[8] is created for that row. When bucket 52 becomes non-zero, the 6th row is initialized to
+     * long[8] and the 5th column is increased to 1. If bucket 53 needs to be incremented, the row already
+     * exists. So the 6th column is incremented.
+     *
+     * In the worst case, this can take storage for 64 longs + 8 row references. But in the normal case, the range
+     * of values for a metric will probably within a 256X range i.e. max value <= min * 256. So at most two rows
+     * will have to be initialized, resulting in 8*2 long number storage + 8 object references.
+     */
+    private final long[][] bucketCounts = new long[8][];
+
+    /**
+     * If count is zero then this field must be zero. Mean can be easily calculated from sum and count.
+     */
+    private double mean;
+
+    private long count;
+
+    private int numBuckets;
+
+    public void add(long value) {
+        int index = findBucketNumber(value);
+        // online calculation of mean and sumOfSquaredDeviation
+        count += 1;
+        double dev = value - mean;
+        mean += dev / count;
+
+        int i = getRow(index);
+        int j = getCol(index);
+        if (bucketCounts[i] == null) {
+            bucketCounts[i] = new long[8];
+        }
+        if (bucketCounts[i][j] == 0) {
+            numBuckets++;
+        }
+        bucketCounts[i][j] += 1;
+    }
+
+    private int findBucketNumber(long value) {
+        if (value < 0) {
+            return 0;
+        }
+        // Algo to find bucket has logarithmic complexity. Max number of
+        // iterations is log_2 64. Note that it is logarithmic wrt number of
+        // buckets and not the value
+        int left = 0;
+        int right = MetricValue.NUM_FINITE_BUCKETS;
+        int mid;
+        while (left != right) {
+            mid = (left + right) / 2;
+            if (value >> mid == 0) {
+                right = mid;
+            } else {
+                left = mid + 1;
+            }
+        }
+        return left + 1;
+    }
+
+    private int getRow(int index) {
+        // index/8
+        return index >> 3;
```

Comment:
You don't need to hand-optimize it, HotSpot will figure it out, see https://stackoverflow.com/questions/1514949/quick-java-optimization-question
Actually it would be great to have this "8" as a constant, if would make more clear what does it mean in different places

Reply:
Thanks for both suggestions. I thought I was being smart with my optimizations. Hotspot is doing almost what I need, so leaving both division and modulo to its optimizations to improve readability. And I am using two constants for ROWS and COLS.